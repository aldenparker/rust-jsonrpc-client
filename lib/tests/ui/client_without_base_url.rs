use jsonrpc_client::{Error, Response, SendRequest, Url};
use serde::de::DeserializeOwned;
use std::fmt;

#[jsonrpc_client::api]
pub trait Math {
    async fn subtract(&self, subtrahend: i64, minuend: i64) -> i64;
}

struct InnerClient;

#[derive(Debug)]
pub struct DummyError;

impl fmt::Display for DummyError {
    fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result {
        unimplemented!()
    }
}

impl std::error::Error for DummyError {}

#[async_trait::async_trait]
impl SendRequest for InnerClient {
    type Error = DummyError;

    async fn send_request<P>(&self, endpoint: Url, body: String) -> Result<Response<P>, Self::Error>
    where
        P: DeserializeOwned,
    {
        unimplemented!()
    }
}

impl From<DummyError> for Error<DummyError> {
    fn from(inner: DummyError) -> Self {
        Self::Client(inner)
    }
}

#[jsonrpc_client::implement(Math)]
pub struct Client {
    inner: InnerClient,
}

fn main() {}
