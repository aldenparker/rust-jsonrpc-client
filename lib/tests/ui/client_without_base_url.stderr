error: struct needs to have either a field named `base_url` or one tagged with `#[jsonrpc_client(base_url)]`
  --> $DIR/client_without_base_url.rs:42:19
   |
42 |   pub struct Client {
   |  ___________________^
43 | |     inner: InnerClient,
44 | | }
   | |_^
